### ISSUE TITLE
Sorting treeview causes crash

### ISSUE THREAD

**Issue by HendrikMennen** [AvaloniaUI/Avalonia#2985]
(Newest nightly)
If I try to sort items in "sub tree" after sorting the "main tree" my application crashes.
It seems that sorting the "main tree" messes up the TreeContainer.

Steps to reproduce:
1. Have a treeview branch with at least one child
2. Remove that treeview branch from the main treeview
3. Add the branch to the main treeview again
4. Remove one child from that branch
5. Add that child to the branch again
6. crash

The crash occured after renaming File3 to File5 in this example
![2019-09-1219-19-15](https://user-images.githubusercontent.com/25281882/64806627-df03fc80-d593-11e9-926d-070c417b0495.gif)

```c#
ProjectFolder top = file.TopFolder;
top.Items.Remove(file);
Console.WriteLine(top.Items.Contains(file)); //returned false
top.Items.Add(file);
```

The last line produces this crash
```
System.ArgumentException: An item with the same key has already been added. Key: VHDPlus.Models.ProjectFiles.ProjectFileVHDP
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at Avalonia.Controls.Generators.TreeContainerIndex.Add(Object item, IControl container) in D:\a\1\s\src\Avalonia.Controls\Generators\TreeContainerIndex.cs:line 52
   at Avalonia.Controls.Generators.TreeItemContainerGenerator`1.CreateContainer(Object item) in D:\a\1\s\src\Avalonia.Controls\Generators\TreeItemContainerGenerator.cs:line 97
   at Avalonia.Controls.Generators.ItemContainerGenerator.Materialize(Int32 index, Object item) in D:\a\1\s\src\Avalonia.Controls\Generators\ItemContainerGenerator.cs:line 59
   at Avalonia.Controls.Presenters.ItemContainerSync.AddContainers(ItemsPresenterBase owner, Int32 index, IEnumerable items) in D:\a\1\s\src\Avalonia.Controls\Presenters\ItemContainerSync.cs:line 91
   at Avalonia.Controls.Presenters.ItemContainerSync.<ItemsChanged>g__Add|0_0(<>c__DisplayClass0_0& ) in D:\a\1\s\src\Avalonia.Controls\Presenters\ItemContainerSync.cs:line 33
   at Avalonia.Controls.Presenters.ItemContainerSync.ItemsChanged(ItemsPresenterBase owner, IEnumerable items, NotifyCollectionChangedEventArgs e) in D:\a\1\s\src\Avalonia.Controls\Presenters\ItemContainerSync.cs:line 43
   at Avalonia.Controls.Presenters.ItemVirtualizerNone.ItemsChanged(IEnumerable items, NotifyCollectionChangedEventArgs e) in D:\a\1\s\src\Avalonia.Controls\Presenters\ItemVirtualizerNone.cs:line 62
   at Avalonia.Reactive.LightweightObservableBase`1.PublishNext(T value) in D:\a\1\s\src\Avalonia.Base\Reactive\LightweightObservableBase.cs:line 132
   at Avalonia.Utilities.WeakSubscriptionManager.Subscription`1.OnEvent(Object sender, T eventArgs) in D:\a\1\s\src\Avalonia.Base\Utilities\WeakSubscriptionManager.cs:line 187
   at System.Collections.ObjectModel.ObservableCollection`1.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
   at System.Collections.ObjectModel.ObservableCollection`1.OnCollectionChanged(NotifyCollectionChangedAction action, Object item, Int32 index)
   at System.Collections.ObjectModel.ObservableCollection`1.InsertItem(Int32 index, T item)
   at VHDPlus.ViewModels.ProjectFilesViewModel.Rename(ProjectEntry file, String oldname) in C:\Users\HendrikMennen\Source\Repos\VHDP\VHDPlus\ViewModels\ProjectFilesViewModel.cs:line 395
   at VHDPlus.Views.ProjectFilesView.Rename_Done(Object sender, RoutedEventArgs e) in C:\Users\HendrikMennen\Source\Repos\VHDP\VHDPlus\Views\ProjectFilesView.xaml.cs:line 364
   at VHDPlus.Views.ProjectFilesView.OnKeyDownHandler(Object sender, KeyEventArgs e) in C:\Users\HendrikMennen\Source\Repos\VHDP\VHDPlus\Views\ProjectFilesView.xaml.cs:line 353
   at Avalonia.Interactivity.Interactive.RaiseEventImpl(RoutedEventArgs e) in D:\a\1\s\src\Avalonia.Interactivity\Interactive.cs:line 209
   at Avalonia.Interactivity.Interactive.BubbleEvent(RoutedEventArgs e) in D:\a\1\s\src\Avalonia.Interactivity\Interactive.cs:line 173
   at Avalonia.Interactivity.Interactive.RaiseEvent(RoutedEventArgs e) in D:\a\1\s\src\Avalonia.Interactivity\Interactive.cs:line 158
   at Avalonia.Input.KeyboardDevice.ProcessRawEvent(RawInputEventArgs e) in D:\a\1\s\src\Avalonia.Input\KeyboardDevice.cs:line 112
   at Avalonia.Input.InputManager.ProcessInput(RawInputEventArgs e) in D:\a\1\s\src\Avalonia.Input\InputManager.cs:line 39
   at Avalonia.Win32.WindowImpl.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam) in D:\a\1\s\src\Windows\Avalonia.Win32\WindowImpl.cs:line 751
   at Avalonia.Win32.Interop.UnmanagedMethods.DispatchMessage(MSG& lpmsg)
   at Avalonia.Win32.Win32Platform.RunLoop(CancellationToken cancellationToken) in D:\a\1\s\src\Windows\Avalonia.Win32\Win32Platform.cs:line 121
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken) in D:\a\1\s\src\Avalonia.Base\Threading\Dispatcher.cs:line 65
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args) in D:\a\1\s\src\Avalonia.Controls\ApplicationLifetimes\ClassicDesktopStyleApplicationLifetime.cs:line 107
   at VHDPlus.Program.Main(String[] args) in C:\Users\HendrikMennen\Source\Repos\VHDP\VHDPlus\Program.cs:line 54
````

**Comment by HendrikMennen** [comment]
Something changed since I created this issue and It got a little harder to cause this crash but it is still possible. I created a repro for it:  https://github.com/HendrikMennen/Avalonia/tree/treeviewcrashrepro

To cause this crash it is important that the `IsExpanded` property is bound to the viewmodel with mode `TwoWay`. 

Here are the steps reproduce this crash every time.
![Unbenannt](https://user-images.githubusercontent.com/25281882/67344038-47e97780-f537-11e9-8ec6-86c41cc2a42e.PNG)
